utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+  }
+  deriving (Show, Generic, FromJSON, ToJSON)
+
+withTextMaybe :: String -> (Text -> Maybe a) -> Value -> Parser a
+withTextMaybe label fromText = withText label $ \text -> case fromText text of
+  Nothing    -> parseFail $ "Unexpected " <> label <> " value: " <> show text
+  Just known -> pure known
+
+data ShaderStage
+  = Vert
+  | Frag
+  | Comp
+  | Tesc
+  | Tese
+  | Geom
+  | Rgen
+  | Rint
+  | Rahit
+  | Rchit
+  | Rmiss
+  | Rcall
+  | Task
+  | Mesh
+  deriving (Eq, Show)
+
+shaderStageFromText :: Text -> Maybe ShaderStage
+shaderStageFromText = \case
+  "vert"  -> Just Vert
+  "frag"  -> Just Frag
+  "comp"  -> Just Comp
+  "tesc"  -> Just Tesc
+  "tese"  -> Just Tese
+  "geom"  -> Just Geom
+  "rgen"  -> Just Rgen
+  "rint"  -> Just Rint
+  "rahit" -> Just Rahit
+  "rchit" -> Just Rchit
+  "rmiss" -> Just Rmiss
+  "rcall" -> Just Rcall
+  "task"  -> Just Task
+  "mesh"  -> Just Mesh
+  _       -> Nothing
+
+shaderStageToText :: ShaderStage -> Text
+shaderStageToText = \case
+  Vert  -> "vert"
+  Frag  -> "frag"
+  Comp  -> "comp"
+  Tesc  -> "tesc"
+  Tese  -> "tese"
+  Geom  -> "geom"
+  Rgen  -> "rgen"
+  Rint  -> "rint"
+  Rahit -> "rahit"
+  Rchit -> "rchit"
+  Rmiss -> "rmiss"
+  Rcall -> "rcall"
+  Task  -> "task"
+  Mesh  -> "mesh"
+
+instance FromJSON ShaderStage where
+  parseJSON = withTextMaybe "mode" shaderStageFromText
+
+instance ToJSON ShaderStage where
+  toJSON     = String . shaderStageToText
+  toEncoding = toEncoding . shaderStageToText
+
+shaderStageFlagBits :: ShaderStage -> ShaderStageFlagBits
+shaderStageFlagBits = \case
+  Vert  -> SHADER_STAGE_VERTEX_BIT
+  Frag  -> SHADER_STAGE_FRAGMENT_BIT
+  Comp  -> SHADER_STAGE_COMPUTE_BIT
+  Tesc  -> SHADER_STAGE_TESSELLATION_CONTROL_BIT
+  Tese  -> SHADER_STAGE_TESSELLATION_EVALUATION_BIT
+  Geom  -> SHADER_STAGE_GEOMETRY_BIT
+  Rgen  -> SHADER_STAGE_RAYGEN_BIT_KHR
+  Rint  -> SHADER_STAGE_INTERSECTION_BIT_KHR
+  Rahit -> SHADER_STAGE_ANY_HIT_BIT_KHR
+  Rchit -> SHADER_STAGE_CLOSEST_HIT_BIT_KHR
+  Rmiss -> SHADER_STAGE_MISS_BIT_KHR
+  Rcall -> SHADER_STAGE_CALLABLE_BIT_KHR
+  Task  -> SHADER_STAGE_TASK_BIT_NV
+  Mesh  -> SHADER_STAGE_MESH_BIT_NV
+
+data Shader = Shader
+  { stage :: ShaderStage
+  , code  :: B.ByteString
+  }
+  deriving Show
+
+data ShaderInfo = ShaderInfo
+  { shaderModuleCreateInfo :: ShaderModuleCreateInfo '[]
+  , pipelineShaderStageCreateInfos
+      :: ShaderModule -> Vector (SomeStruct PipelineShaderStageCreateInfo)
+  }
+
+makeShaderInfo :: (Shader, Reflection) -> ShaderInfo
+makeShaderInfo (Shader {..}, Reflection {..}) = ShaderInfo { .. }
+ where
+  shaderModuleCreateInfo = makeShaderModuleCreateInfo code
+  pipelineShaderStageCreateInfos =
+    (SomeStruct <$>) . makePipelineShaderStageCreateInfos stage entryPoints
+
+makeDescriptorInfo
+  :: Vector (Shader, Reflection) -> Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorInfo = makeDescriptorSetLayoutCreateInfos . join . V.map
+  (   makeDescriptorSetLayoutBindings
```

Comment:
This pipeline is kinda confusing.

I expect it to be significantly more legible if you turn it into a local function operating on a single `(Shader, Reflection)` element.

Reply:
fixed with
``` haskell
makeDescriptorSetLayoutBindings :: ShaderStage -> Vector Ubo -> Vector Texture -> Vector (Int, DescriptorSetLayoutBinding)

makeDescriptorInfo :: Vector (Shader, Reflection) -> Vector (DescriptorSetLayoutCreateInfo '[])
makeDescriptorInfo = makeDescriptorSetLayoutCreateInfos . join . V.map
  (   makeDescriptorSetLayoutBindings
  <$> makeShaderStage
  <*> makeUbos
  <*> makeTextures
  )
 where
  makeShaderStage :: (Shader, Reflection) -> ShaderStage
  makeShaderStage = (stage :: Shader -> ShaderStage) . fst
  makeUbos :: (Shader, Reflection) -> Vector Ubo
  makeUbos = fromMaybe [] . ubos . snd
  makeTextures :: (Shader, Reflection) -> Vector Texture
  makeTextures = fromMaybe [] . textures . snd
```

Reply:
Do-notation for vectors should look really good here (if this works 😅)

```haskell
makeDescriptorInfo shaders = makeDescriptorSetLayoutCreateInfos $ do
  (Shader{stage}, Reflection{ubos, textures}) <- shaders
  makeDescriptorSetLayoutBindings stage (fromMaybe [] ubos) (fromMaybe [] textures)
```
Pattern matching gets rid of the extra functions and no `join` needed since the output is a `Vector` too.

Reply:
only the `join` in this place has been replace to do-notation style. 

there are some `join` usage in other place but I think they look well